File: LinkedList/19RemoveNthNode/one.js

// * Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {

    let last = null
    let cur = head

    while(cur){
        const next = cur.next
        cur.next = last
        last = cur
        cur = next  
    }

    head = last

    removeNth(head,n)
    showAll(head)

    function removeNth(head,n){
        let pre
        cur = head
        for(let i = 1; i<n;i++){
            pre = cur
            cur = cur.next
        }

        
        if(pre)
            if(cur)
                pre.next = cur.next 
        else{
            pre = cur
            cur = cur.next
            pre.next = cur.next
        }
    }
};

function showAll(head){
    let tmp = head
    while(tmp){
        console.log(tmp.val)
        tmp = tmp.next
    }
}

// let head = new ListNode(1,new ListNode(2,new ListNode(3,new ListNode(4,new ListNode(5,null)))))
let head = new ListNode(1,new ListNode(2,null))

let n = 1

removeNthFromEnd(head,n)

File: LinkedList/19RemoveNthNode/one.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def print(self):
        cur = self
        while(cur):
            print(cur.val)
            cur = cur.next

class Solution(object):
    def removeNthFromEnd(self, head, n):

        def getLength(head):
            length = 0
            while(head):
                length += 1
                head = head.next
            return length
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        quick = head
        slow = head

        length = getLength(head)

        for i in range (n):
            quick = quick.next


        for i in range(n,length-1):
            quick = quick.next
            slow = slow.next

        slow.next = quick

        head.print()
        return head


        
    def main(self):
        # head = ListNode(1,ListNode(2,ListNode(3,ListNode(4,ListNode(5,None)))))
        head = ListNode(1,None)
        n = 1
        self.removeNthFromEnd(head,n)


myInstance = Solution()

myInstance.main()